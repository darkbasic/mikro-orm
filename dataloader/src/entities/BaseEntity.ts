import {IdEntity, PrimaryKey} from 'mikro-orm';

interface BaseEntityConstructor {
  id?: number;
}

export abstract class BaseEntity implements IdEntity<BaseEntity> {
  @PrimaryKey()
  id!: number;

  constructor({id}: BaseEntityConstructor = {}) {
    if (id) {
      this.id = id;
    }
  }
}
