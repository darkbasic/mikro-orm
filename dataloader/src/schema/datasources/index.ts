export * from './OrmAPI';
export * from './UserAPI';
export * from './SportAPI';
export * from './SpecialtyAPI';
export * from './SiteAPI';
export * from './MatchAPI';
