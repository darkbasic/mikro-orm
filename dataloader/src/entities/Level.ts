import {
  Entity,
  Property,
  ManyToOne,
  Unique,
  IdentifiedReference,
  Reference,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {User} from './User';
import {Sport} from './Sport';

interface LevelConstructor {
  user: User;
  sport: Sport;
  declaredLevel: number;
}

export {Level as LevelEntity};

@Entity()
@Unique({properties: ['user', 'sport']})
export class Level extends BaseEntity {
  @ManyToOne()
  user: IdentifiedReference<User>;

  @ManyToOne()
  sport: IdentifiedReference<Sport>;

  @Property({type: 'float'})
  declaredLevel: number;

  constructor({user, sport, declaredLevel}: LevelConstructor) {
    super();
    this.user = Reference.create(user);
    this.sport = Reference.create(sport);
    this.declaredLevel = declaredLevel;
  }
}
