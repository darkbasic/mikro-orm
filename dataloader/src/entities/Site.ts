import {Entity, Property, ManyToMany, Collection, OneToMany} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {Sport} from './Sport';
import {Match} from './Match';

interface SiteConstructor {
  id?: number;
  name: string;
  position: string;
  sports: Sport[];
}

export {Site as SiteEntity};

@Entity()
export class Site extends BaseEntity {
  @Property()
  name: string;

  @Property()
  position: string;

  @ManyToMany()
  sports: Collection<Sport> = new Collection<Sport>(this);

  @OneToMany(() => Match, match => match.site)
  matches: Collection<Match> = new Collection<Match>(this);

  constructor({id, name, position, sports}: SiteConstructor) {
    super({id});
    this.name = name;
    this.position = position;
    this.sports.add(...sports);
  }
}
