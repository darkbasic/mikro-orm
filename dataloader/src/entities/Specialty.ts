import {
  Entity,
  Property,
  ManyToOne,
  OneToMany,
  Collection,
  IdentifiedReference,
  Reference,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {Sport} from './Sport';
import {Match} from './Match';

interface SpecialtyConstructor {
  id?: number;
  name: string;
  males: number;
  females: number;
  sport: Sport;
}

export {Specialty as SpecialtyEntity};

@Entity()
export class Specialty extends BaseEntity {
  @Property()
  name: string;

  @Property()
  males: number;

  @Property()
  females: number;

  @ManyToOne()
  sport: IdentifiedReference<Sport>;

  @OneToMany(() => Match, match => match.specialty)
  matches: Collection<Match> = new Collection<Match>(this);

  constructor({id, name, males, females, sport}: SpecialtyConstructor) {
    super({id});
    this.name = name;
    this.males = males;
    this.females = females;
    this.sport = Reference.create(sport);
  }
}
