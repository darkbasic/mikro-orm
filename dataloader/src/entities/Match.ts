import {
  Entity,
  Property,
  ManyToMany,
  Collection,
  ManyToOne,
  IdentifiedReference,
  Reference,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {User} from './User';
import {Specialty} from './Specialty';
import {Site} from './Site';

interface MatchConstructor {
  date: Date;
  creator: User;
  specialty: Specialty;
  site: Site;
  partecipants: User[];
}

export {Match as MatchEntity};

@Entity()
export class Match extends BaseEntity {
  @Property()
  date: Date;

  @ManyToOne()
  creator: IdentifiedReference<User>;

  @ManyToOne()
  specialty: IdentifiedReference<Specialty>;

  @ManyToOne()
  site: IdentifiedReference<Site>;

  @ManyToMany()
  partecipants: Collection<User> = new Collection<User>(this);

  constructor({
    date,
    creator,
    specialty,
    site,
    partecipants,
  }: MatchConstructor) {
    super();
    this.date = date;
    this.creator = Reference.create(creator);
    this.specialty = Reference.create(specialty);
    this.site = Reference.create(site);
    this.partecipants.add(...partecipants);
  }
}
