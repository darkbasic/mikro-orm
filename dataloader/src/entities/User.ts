import {
  Entity,
  Property,
  Enum,
  OneToMany,
  Collection,
  ManyToMany,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {Match} from './Match';

export enum Sex {
  MALE,
  FEMALE,
}

interface UserContructor {
  id?: number;
  email: string;
  name: string;
  surname: string;
  sex: Sex;
  password: string;
  picture?: string;
}

export {User as UserEntity};

@Entity()
export class User extends BaseEntity {
  @Property()
  email: string;

  @Property()
  name: string;

  @Property()
  surname: string;

  @Enum()
  sex: Sex;

  @Property()
  password: string;

  @Property()
  picture?: string;

  @OneToMany(() => Match, match => match.creator)
  createdMatches: Collection<Match> = new Collection<Match>(this);

  @ManyToMany(() => Match, match => match.partecipants)
  partecipatedMatches: Collection<Match> = new Collection<Match>(this);

  constructor({
    id,
    email,
    name,
    surname,
    sex,
    password,
    picture,
  }: UserContructor) {
    super({id});
    this.email = email;
    this.name = name;
    this.surname = surname;
    this.sex = sex;
    this.password = password;
    if (picture) {
      this.picture = picture;
    }
  }
}
