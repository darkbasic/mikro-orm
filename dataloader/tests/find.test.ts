import {MikroORM, EntityRepository} from 'mikro-orm';
import {PostgreSqlDriver} from 'mikro-orm/dist/drivers/PostgreSqlDriver';
import {initORMPostgreSql, resetDatabase} from './bootstrap';
import {User} from '../src/entities/User';
import {EntityDataLoader} from '../src/schema/datasources/utils';
import {Level} from '../src/entities/Level';
import {Match} from '../src/entities/Match';

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe('Backend', () => {
  jest.setTimeout(10e3);
  let orm: MikroORM<PostgreSqlDriver>;
  let userRepo: EntityRepository<User>;
  let levelRepo: EntityRepository<Level>;
  let matchRepo: EntityRepository<Match>;
  let dataloader: EntityDataLoader;

  beforeAll(async () => {
    orm = await initORMPostgreSql();
    await resetDatabase(orm);
    userRepo = orm.em.getRepository(User);
    levelRepo = orm.em.getRepository(Level);
    matchRepo = orm.em.getRepository(Match);
    dataloader = new EntityDataLoader(orm.em);
  });

  beforeEach(async () => {
    orm.em.clear();
    await sleep(100);
  });

  test('isConnected()', async () => {
    await expect(orm.isConnected()).resolves.toBe(true);
    await orm.close(true);
    await expect(orm.isConnected()).resolves.toBe(false);
    await orm.connect();
    await expect(orm.isConnected()).resolves.toBe(true);
  });

  it('should retrieve parallel find queries with EntityDataLoader', async () => {
    const levelsDataloader = await Promise.all(
      [1, 3, 4, 5, 999].map(declaredLevel =>
        dataloader.find(levelRepo, {declaredLevel})
      )
    );
    expect(levelsDataloader).toBeDefined();
    expect(levelsDataloader.length).toEqual(5);
    expect(levelsDataloader).toMatchSnapshot();
  });

  it('should return the same entities without EntityDataLoader', async () => {
    const conditions = [1, 3, 4, 5, 999];
    const levels = await Promise.all(
      conditions.map(declaredLevel => levelRepo.find({declaredLevel}))
    );
    const levelsDataloader = await Promise.all(
      conditions.map(declaredLevel =>
        dataloader.find(levelRepo, {declaredLevel})
      )
    );
    expect(JSON.stringify(levels)).toBe(JSON.stringify(levelsDataloader));
  });

  it('should be faster with EntityDataLoader', async () => {
    const conditions = [1, 3, 4, 5, 999];
    const start = new Date();
    await Promise.all(
      conditions.map(declaredLevel => levelRepo.find({declaredLevel}))
    );
    const end = new Date();
    const diff = end.getTime() - start.getTime();
    const startDataloader = new Date();
    await Promise.all(
      conditions.map(declaredLevel =>
        dataloader.find(levelRepo, {declaredLevel})
      )
    );
    const endDataloader = new Date();
    const diffDataloader = endDataloader.getTime() - startDataloader.getTime();
    console.log(`W/o dataloader: ${diff} ms`);
    console.log(`W/ dataloader: ${diffDataloader} ms`);
    expect(diffDataloader).toBeLessThan(diff);
  });

  it('should retrieve parallel find queries with conditions on reference fields with EntityDataLoader', async () => {
    const matchesDataloader = await Promise.all(
      [
        {specialty: 1, site: 1},
        {specialty: 1, site: 2},
        {specialty: 1, site: [1, 2]},
        {specialty: [1, 2], site: [1, 2]},
        {specialty: [3], site: 1},
        {specialty: [1, 2, 3], site: [1, 999]},
      ].map(({specialty, site}) =>
        dataloader.find(matchRepo, {specialty, site})
      )
    );
    expect(matchesDataloader).toBeDefined();
    expect(matchesDataloader.length).toEqual(6);
    expect(matchesDataloader).toMatchSnapshot();
  });

  it('should return the same entities without EntityDataLoader', async () => {
    const conditions = [
      {specialty: 1, site: 1},
      {specialty: 1, site: 2},
      {specialty: 1, site: [1, 2]},
      {specialty: [1, 2], site: [1, 2]},
      {specialty: [3], site: 1},
      {specialty: [1, 2, 3], site: [1, 999]},
    ];
    const matches = await Promise.all(
      conditions.map(({specialty, site}) => matchRepo.find({specialty, site}))
    );
    const matchesDataloader = await Promise.all(
      conditions.map(({specialty, site}) =>
        dataloader.find(matchRepo, {specialty, site})
      )
    );
    expect(JSON.stringify(matches)).toBe(JSON.stringify(matchesDataloader));
  });

  it('should be faster with EntityDataLoader', async () => {
    const conditions = [
      {specialty: 1, site: 1},
      {specialty: 1, site: 2},
      {specialty: 1, site: [1, 2]},
      {specialty: [1, 2], site: [1, 2]},
      {specialty: [3], site: 1},
      {specialty: [1, 2, 3], site: [1, 999]},
    ];
    const start = new Date();
    await Promise.all(
      conditions.map(({specialty, site}) => matchRepo.find({specialty, site}))
    );
    const end = new Date();
    const diff = end.getTime() - start.getTime();
    const startDataloader = new Date();
    await Promise.all(
      conditions.map(({specialty, site}) =>
        dataloader.find(matchRepo, {specialty, site})
      )
    );
    const endDataloader = new Date();
    const diffDataloader = endDataloader.getTime() - startDataloader.getTime();
    console.log(`W/o dataloader: ${diff} ms`);
    console.log(`W/ dataloader: ${diffDataloader} ms`);
    expect(diffDataloader).toBeLessThan(diff);
  });

  it('should retrieve parallel find queries with conditions on collection fields with EntityDataLoader', async () => {
    const partecipantsDataloader = await Promise.all(
      [1, 2, 3, 4, 5].map(partecipatedMatches =>
        dataloader.find(userRepo, {partecipatedMatches})
      )
    );
    expect(partecipantsDataloader).toBeDefined();
    expect(partecipantsDataloader.length).toEqual(5);
    expect(partecipantsDataloader).toMatchSnapshot();
  });

  it('should return the same entities without EntityDataLoader', async () => {
    const partecipants = await Promise.all(
      [1, 2, 3, 4, 5].map(partecipatedMatches =>
        userRepo.find({partecipatedMatches})
      )
    );
    const partecipantsDataloader = await Promise.all(
      [1, 2, 3, 4, 5].map(partecipatedMatches =>
        dataloader.find(userRepo, {partecipatedMatches})
      )
    );
    expect(
      JSON.stringify(
        partecipants.map(users => users.sort((a, b) => a.id - b.id))
      )
    ).toBe(
      JSON.stringify(
        partecipantsDataloader.map(users => users.sort((a, b) => a.id - b.id))
      )
    );
  });

  it('should be faster with EntityDataLoader', async () => {
    const start = new Date();
    await Promise.all(
      [1, 2, 3, 4, 5].map(partecipatedMatches =>
        userRepo.find({partecipatedMatches})
      )
    );
    const end = new Date();
    const diff = end.getTime() - start.getTime();
    const startDataloader = new Date();
    await Promise.all(
      [1, 2, 3, 4, 5].map(partecipatedMatches =>
        dataloader.find(userRepo, {partecipatedMatches})
      )
    );
    const endDataloader = new Date();
    const diffDataloader = endDataloader.getTime() - startDataloader.getTime();
    console.log(`W/o dataloader: ${diff} ms`);
    console.log(`W/ dataloader: ${diffDataloader} ms`);
    expect(diffDataloader).toBeLessThan(diff);
  });

  afterAll(async () => {
    orm.close(true);
    await sleep(100);
  });
});
