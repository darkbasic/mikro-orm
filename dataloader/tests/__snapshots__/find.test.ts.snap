// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend should retrieve parallel find queries with EntityDataLoader 1`] = `
Array [
  Array [
    Object {
      "declaredLevel": 1,
      "id": 5,
      "sport": 1,
      "user": 7,
    },
    Object {
      "declaredLevel": 1,
      "id": 7,
      "sport": 2,
      "user": 2,
    },
    Object {
      "declaredLevel": 1,
      "id": 11,
      "sport": 1,
      "user": 8,
    },
    Object {
      "declaredLevel": 1,
      "id": 12,
      "sport": 2,
      "user": 10,
    },
  ],
  Array [
    Object {
      "declaredLevel": 3,
      "id": 2,
      "sport": 1,
      "user": 2,
    },
  ],
  Array [
    Object {
      "declaredLevel": 4,
      "id": 4,
      "sport": 1,
      "user": 6,
    },
  ],
  Array [
    Object {
      "declaredLevel": 5,
      "id": 1,
      "sport": 1,
      "user": 1,
    },
    Object {
      "declaredLevel": 5,
      "id": 8,
      "sport": 2,
      "user": 3,
    },
  ],
  Array [],
]
`;

exports[`Backend should retrieve parallel find queries with conditions on collection fields with EntityDataLoader 1`] = `
Array [
  Array [
    Object {
      "email": "niccolo.belli@linuxsystems.it",
      "id": 1,
      "name": "Niccolò",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "1111",
      "picture": null,
      "sex": 0,
      "surname": "Belli",
    },
    Object {
      "email": "mario.rossi@linuxsystems.it",
      "id": 2,
      "name": "Mario",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "2222",
      "picture": null,
      "sex": 0,
      "surname": "Rossi",
    },
    Object {
      "email": "mattia.ferrari@linuxsystems.it",
      "id": 4,
      "name": "Mattia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "4444",
      "picture": null,
      "sex": 0,
      "surname": "Ferrari",
    },
    Object {
      "email": "filippo.magi@linuxsystems.it",
      "id": 6,
      "name": "Filippo",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
      ],
      "password": "6666",
      "picture": null,
      "sex": 0,
      "surname": "Magi",
    },
  ],
  Array [
    Object {
      "email": "niccolo.belli@linuxsystems.it",
      "id": 1,
      "name": "Niccolò",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "1111",
      "picture": null,
      "sex": 0,
      "surname": "Belli",
    },
    Object {
      "email": "mattia.ferrari@linuxsystems.it",
      "id": 4,
      "name": "Mattia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "4444",
      "picture": null,
      "sex": 0,
      "surname": "Ferrari",
    },
    Object {
      "email": "sara.ricci@linuxsystems.it",
      "id": 7,
      "name": "Sara",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "7777",
      "picture": null,
      "sex": 1,
      "surname": "Ricci",
    },
    Object {
      "email": "giorgia.esposito@linuxsystems.it",
      "id": 9,
      "name": "Giorgia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "9999",
      "picture": null,
      "sex": 1,
      "surname": "Esposito",
    },
  ],
  Array [
    Object {
      "email": "niccolo.belli@linuxsystems.it",
      "id": 1,
      "name": "Niccolò",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "1111",
      "picture": null,
      "sex": 0,
      "surname": "Belli",
    },
    Object {
      "email": "mario.rossi@linuxsystems.it",
      "id": 2,
      "name": "Mario",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "2222",
      "picture": null,
      "sex": 0,
      "surname": "Rossi",
    },
    Object {
      "email": "mattia.ferrari@linuxsystems.it",
      "id": 4,
      "name": "Mattia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "4444",
      "picture": null,
      "sex": 0,
      "surname": "Ferrari",
    },
    Object {
      "email": "luigi.bianchi@linuxsystems.it",
      "id": 3,
      "name": "Luigi",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "3333",
      "picture": null,
      "sex": 0,
      "surname": "Bianchi",
    },
  ],
  Array [
    Object {
      "email": "mattia.ferrari@linuxsystems.it",
      "id": 4,
      "name": "Mattia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "4444",
      "picture": null,
      "sex": 0,
      "surname": "Ferrari",
    },
    Object {
      "email": "filippo.magi@linuxsystems.it",
      "id": 6,
      "name": "Filippo",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
      ],
      "password": "6666",
      "picture": null,
      "sex": 0,
      "surname": "Magi",
    },
    Object {
      "email": "giorgia.esposito@linuxsystems.it",
      "id": 9,
      "name": "Giorgia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "9999",
      "picture": null,
      "sex": 1,
      "surname": "Esposito",
    },
    Object {
      "email": "claudia.recanatini@linuxsystems.it",
      "id": 10,
      "name": "Claudia",
      "partecipatedMatches": Array [
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "0000",
      "picture": null,
      "sex": 1,
      "surname": "Recanatini",
    },
  ],
  Array [
    Object {
      "email": "sara.ricci@linuxsystems.it",
      "id": 7,
      "name": "Sara",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "7777",
      "picture": null,
      "sex": 1,
      "surname": "Ricci",
    },
    Object {
      "email": "giorgia.esposito@linuxsystems.it",
      "id": 9,
      "name": "Giorgia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "9999",
      "picture": null,
      "sex": 1,
      "surname": "Esposito",
    },
    Object {
      "email": "claudia.recanatini@linuxsystems.it",
      "id": 10,
      "name": "Claudia",
      "partecipatedMatches": Array [
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "0000",
      "picture": null,
      "sex": 1,
      "surname": "Recanatini",
    },
    Object {
      "email": "lucia.fontana@linuxsystems.it",
      "id": 8,
      "name": "Lucia",
      "partecipatedMatches": Array [
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "8888",
      "picture": null,
      "sex": 1,
      "surname": "Fontana",
    },
  ],
]
`;

exports[`Backend should retrieve parallel find queries with conditions on reference fields with EntityDataLoader 1`] = `
Array [
  Array [
    Object {
      "creator": 1,
      "date": 2020-03-12T18:00:00.000Z,
      "id": 1,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 0,
        "id": 1,
        "males": 4,
        "name": "2x2 Male",
        "sport": 1,
      },
    },
    Object {
      "creator": 1,
      "date": 2020-03-15T18:00:00.000Z,
      "id": 3,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 0,
        "id": 1,
        "males": 4,
        "name": "2x2 Male",
        "sport": 1,
      },
    },
  ],
  Array [],
  Array [
    Object {
      "creator": 1,
      "date": 2020-03-12T18:00:00.000Z,
      "id": 1,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 0,
        "id": 1,
        "males": 4,
        "name": "2x2 Male",
        "sport": 1,
      },
    },
    Object {
      "creator": 1,
      "date": 2020-03-15T18:00:00.000Z,
      "id": 3,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 0,
        "id": 1,
        "males": 4,
        "name": "2x2 Male",
        "sport": 1,
      },
    },
  ],
  Array [
    Object {
      "creator": 1,
      "date": 2020-03-12T18:00:00.000Z,
      "id": 1,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 0,
        "id": 1,
        "males": 4,
        "name": "2x2 Male",
        "sport": 1,
      },
    },
    Object {
      "creator": 7,
      "date": 2020-03-15T18:00:00.000Z,
      "id": 5,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 4,
        "id": 2,
        "males": 0,
        "name": "2x2 Female",
        "sport": 1,
      },
    },
    Object {
      "creator": 1,
      "date": 2020-03-15T18:00:00.000Z,
      "id": 3,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 0,
        "id": 1,
        "males": 4,
        "name": "2x2 Male",
        "sport": 1,
      },
    },
  ],
  Array [
    Object {
      "creator": 1,
      "date": 2020-05-20T16:30:00.000Z,
      "id": 2,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 2,
        "id": 3,
        "males": 2,
        "name": "2x2 Mixed",
        "sport": 1,
      },
    },
  ],
  Array [
    Object {
      "creator": 1,
      "date": 2020-03-12T18:00:00.000Z,
      "id": 1,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 0,
        "id": 1,
        "males": 4,
        "name": "2x2 Male",
        "sport": 1,
      },
    },
    Object {
      "creator": 1,
      "date": 2020-05-20T16:30:00.000Z,
      "id": 2,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 2,
        "id": 3,
        "males": 2,
        "name": "2x2 Mixed",
        "sport": 1,
      },
    },
    Object {
      "creator": 7,
      "date": 2020-03-15T18:00:00.000Z,
      "id": 5,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 4,
        "id": 2,
        "males": 0,
        "name": "2x2 Female",
        "sport": 1,
      },
    },
    Object {
      "creator": 1,
      "date": 2020-03-15T18:00:00.000Z,
      "id": 3,
      "site": Object {
        "id": 1,
        "name": "The Beach",
        "position": "Falconara Marittima",
      },
      "specialty": Object {
        "females": 0,
        "id": 1,
        "males": 4,
        "name": "2x2 Male",
        "sport": 1,
      },
    },
  ],
]
`;
